subscription-cache/src/lib.rs

```diff
@@ -0,0 +1,130 @@
+    }
+}
+
+impl Hash for SubscriptionInformation {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.subscriber.hash(state);
+    }
+}
+
+impl Clone for SubscriptionInformation {
+    fn clone(&self) -> Self {
+        Self {
+            subscriber: self.subscriber.clone(),
+            status: self.status.clone(),
+            attributes: self.attributes.clone(),
+            config: self.config.clone(),
+        }
+    }
+}
+
+pub struct SubscriptionCache {
+    subscription_cache_map: SubscribersMap,
+}
+
+/// A [`SubscriptionCache`] is used to store and manage subscriptions to
+/// topics. It is kept local to the streamer. The streamer will receive updates
+/// from the subscription service, and update the SubscriptionCache accordingly.
+impl SubscriptionCache {
+    pub fn new(subscription_cache_map: FetchSubscriptionsResponse) -> Self {
+        let mut subscription_cache_hash_map = HashMap::new();
+        for subscription in subscription_cache_map.subscriptions {
+            let uri = if let Some(uri) = subscription.topic.into_option(){
+                uri
+            } else {
+                println!("Unable to parse URI from subscription, skipping...");
+                continue;
+            };
+            let subscriber = if let Some(subscriber) = subscription.subscriber.into_option(){
+                subscriber
+            } else {
+                println!("Unable to parse subscriber from subscription, skipping...");
+                continue;
+            };
+            let status = if let Some(status) = subscription.status.into_option(){
+                status
+            } else {
+                println!("Unable to parse status from subscription, setting as default");
+                SubscriptionStatus::default()
+            };
+            let attributes = if let Some(attributes) = subscription.attributes.into_option(){
+                attributes
+            } else {
+                println!("Unable to parse attributes from subscription, setting as default");
+                SubscribeAttributes::default()
+            };
+            let config = if let Some(config) = subscription.config.into_option(){
+                config
+            } else {
+                println!("Unable to parse config from subscription, setting as default");
+                EventDeliveryConfig::default()
+            };
+            // Create new hashset if the key does not exist and insert the subscription
+            let subscription_information = SubscriptionInformation {
+                subscriber,
+                status,
+                attributes,
+                config
+            };
+            subscription_cache_hash_map
+                .entry(uri)
+                .or_insert_with(HashSet::new)
+                .insert(subscription_information);
+        }
+        Self {
+            subscription_cache_map: Mutex::new(subscription_cache_hash_map),
+        }
+    }
+
+    pub async fn fetch_cache(
```

Comment:
Feels like this could be improved from an efficiency stand-point.

Some questions to ask ourselves:
1. Do we need to make a clone of the whole `subscription_cache_map`?
2. Is there a way to write a function on the `SubscriptionCache` that will let us query directly for the `authority_name` we need?
3. Is there a way to store the information in the `SubscriptionCache` which will facilitate fast lookups based on the `authority_name`?

Reply:
I think I addressed number 3 here so far, need to see how to deal with 1. and 2.. 

2 seems easier, so will try that next

Reply:
well fetch_cache was causing a clippy issue, and honestly I don't know/remember why I was cloning the map like that... I found a much cleaner way to return the cache. let me try to do it better now